src/annotator/sidebar.js

```diff
@@ -24,14 +24,130 @@ const defaultConfig = {
   },
 };
 
-export default class Sidebar extends Host {
+/**
+ * Create the JSON-serializable subset of annotator configuration that should
+ * be passed to the sidebar application.
+ */
+function createSidebarConfig(config) {
+  const sidebarConfig = { ...config };
+
+  // Some config settings are not JSON-stringifiable (e.g. JavaScript
+  // functions) and will be omitted when the config is JSON-stringified.
+  // Add a JSON-stringifiable option for each of these so that the sidebar can
+  // at least know whether the callback functions were provided or not.
+  if (sidebarConfig.services?.length > 0) {
+    const service = sidebarConfig.services[0];
+    if (service.onLoginRequest) {
+      service.onLoginRequestProvided = true;
+    }
+    if (service.onLogoutRequest) {
+      service.onLogoutRequestProvided = true;
+    }
+    if (service.onSignupRequest) {
+      service.onSignupRequestProvided = true;
+    }
+    if (service.onProfileRequest) {
+      service.onProfileRequestProvided = true;
+    }
+    if (service.onHelpRequest) {
+      service.onHelpRequestProvided = true;
+    }
+  }
+
+  // Remove several annotator-only properties.
+  //
+  // nb. We don't currently strip all the annotator-only properties here.
+  // That's OK because validation / filtering happens in the sidebar app itself.
+  // It just results in unnecessary content in the sidebar iframe's URL string.
+  ['sidebarAppUrl', 'pluginClasses'].forEach(key => delete sidebarConfig[key]);
+
+  return sidebarConfig;
+}
+
+/**
+ * Create the iframe that will load the sidebar application.
+ *
+ * @return {HTMLIFrameElement}
+ */
+function createSidebarIframe(config) {
+  const sidebarConfig = createSidebarConfig(config);
+  const configParam =
+    'config=' + encodeURIComponent(JSON.stringify(sidebarConfig));
+  const sidebarAppSrc = config.sidebarAppUrl + '#' + configParam;
+
+  const sidebarFrame = document.createElement('iframe');
+  sidebarFrame.setAttribute('name', 'hyp_sidebar_frame');
+
+  // Enable media in annotations to be shown fullscreen
+  sidebarFrame.setAttribute('allowfullscreen', '');
+
+  sidebarFrame.setAttribute('seamless', '');
+  sidebarFrame.src = sidebarAppSrc;
+  sidebarFrame.title = 'Hypothesis annotation viewer';
+  sidebarFrame.className = 'h-sidebar-iframe';
+
+  return sidebarFrame;
+}
+
+/**
+ * The `Sidebar` class creates the sidebar application iframe and its container,
+ * as well as the controls that are displayed adjacent to it.
+ */
+export default class Sidebar extends Guest {
   constructor(element, config) {
     if (config.theme === 'clean' || config.externalContainerSelector) {
       delete config.pluginClasses.BucketBar;
     }
 
+    let externalContainer = null;
+
+    if (config.externalContainerSelector) {
+      externalContainer = document.querySelector(
```

Comment:
I've left this behaviour as it was in the previous code, but I agree it would make sense to warn. I don't know if anyone is actually still using `externalContainerSelector`. We might be able to remove this entirely.